### ISSUE TITLE
A Bayesian Perspective on Generalization and Stochastic Gradient Descent

### ISSUE THREAD

**Issue by yoheikikuta** [yoheikikuta/paper-reading#5]
## 論文リンク
https://arxiv.org/abs/1710.06451

## 公開日（yyyy/mm/dd）
2017/10/17

## 概要
SGD による学習で DL model がどのような汎化性能を示すかに関して、ベイジアンの観点から調べた論文。

まずは Bayesian evidence を考えることで従来言われていたような broad minima の方が sharp minima よりも汎化性能が高いことを自然に説明する。

SGD を連続化して確率微分方程式としてみなし、full batch と mini batch におけるパラメタ更新の差分をランジュバン方程式のノイズと対応させることで、そのノイズのスケールを g ~ ε N / B (ε: learning rate, N: training set size, B: batch size) と表せることを示す。

この noise スケールが shape minima から遠ざける働きを担うもので、適切な量に設定することで良い汎化性能を持つ解にたどり着けるという話。特に学習率とバッチサイズの比を保つことで汎化性能一定の解が得られるというのが昨今重要な発見であり、それを理論的にサポートしている。

**Comment by yoheikikuta** [comment]
この論文に至るまで、解の近傍で曲率が小さくなるような broad minima の方が汎化性能が高い解ではないかという話が議論されていた。PAC bayesian の議論で sharp minima に penalty を与えるという研究もある。これは以前読んだことある論文で https://arxiv.org/abs/1706.08947 ですな。

この論文ではどうやって汎化性能測るのかということを議論していたもので、従来の VC-dim とか Rademacher complexity が DL に適用し難いというところが重要なスタートポイントになっている。例えば DL はランダムラベルにフィットできてしまう。
そこで重み行列の norm を調べたり、パラメタを変更したときの誤差の変化分に関して PAC bayes の手法で調べたりすることで、汎化性能との関係を調べている。

上述のように sharpness との対応などは発見されたが、実験的に色々試してみたという感じで、理論的に「これが汎化性能を測るのに使える指標だ」と一意に決まるような結果は得られていないという理解。

**Comment by yoheikikuta** [comment]
bayesian の議論を始める上で、パラメタの事後確率から始める。

<p align="center">
  <img src="https://imgur.com/4rUS72s.png">
</p>

ここで、尤度に関して exponentiate して prior として gaussian を使うことを考えれば、分子は e^{-C} の形で書けるので、新しいラベルの予測に関しては以下の式で書ける。

<p align="center">
  <img src="https://imgur.com/Gh8BIG1.png">
</p>

分母は規格化のために形式的に書かれている積分である。
2つのモデル M1, M2 のどちらが良いモデルかを比較するには以下で定まる。

<p align="center">
  <img src="https://imgur.com/R8n39id.png">
</p>

モデルの prior というのは決めるのが難しいので、ここでは P(M1) = P(M2)  として、主たる解析対象は p(y|x; M) とする。一般論は展開するのが困難なので、パラメタとしてある種の最適解の周りだけ考えることにして（そこで確率が dominate するとして）議論を展開する。

**Comment by yoheikikuta** [comment]
パラメタをある点 w0 において taylor 展開して二次まで保持すると以下が得られる。
まあまあ省略してしまっているが、 $` p(y|x;M) = \int \ dw \ p(y|w,x;M) p(w;M) `$ で prior を gaussian にしている。さらに簡単のためにパラメタは一次元のパラメタであるとしている。

<p align="center">
  <img src="https://imgur.com/1vA3s7L.png">
</p>

これはラプラス展開になっているわけだが、これから evidence ( $` p(y|x;M) `$ のことをそう呼ぶ) が解における曲率と正則化項の係数との比の log でコントロールされることが分かる。

**Comment by yoheikikuta** [comment]
パラメタ数を複数に一般化するのは難しくない。Hessian の determinant を計算することになるが、固有値の積で書けることに注意すれば以下が得られる。

<p align="center">
  <img src="https://imgur.com/saxJhjs.png">
</p>

この Hessian の項は Occam factor と呼ばれるとのこと。まあ単純にモデル比較においてこの項の寄与が小さい（ざっくり言ってしまってパラメタがシンプルな方がこの項は小さくなる）方が選ばれやすいというのが理由である。

この evidence をこの論文では null model と比べる。これは全てのラベルを random としてどのクラスに assign するかは等確率で定められるというものである。この場合パラメタは存在せず、evidence は $` (1/n)^N `$ (n: クラス数, N: 学習データ数) となる。よって

<p align="center">
  <img src="https://imgur.com/vzLxmDk.png">
</p>

$` E = C + \frac{1}{2} \sum \ \ln(λi / λ) - N \ \ln(n) `$ であり、これが 0 未満のときにモデルが予測に対して何かしらの confidence を与えることになる。

ここから、broad minima が shape minima よりもよく汎化する解を与えるというこれまで知られていた事実をサポートしている。基本的な bayesian の式変形と解の周りの taylor 展開を使っているだけだが、興味深い。

**Comment by yoheikikuta** [comment]
実験: 意味のある情報で学習したモデルは良く一般化できることを示す

- DL 特有の話でないことを示すため、logistic regression
- 0,1 のラベルだけを使う MNIST
- training は 400+400 images
- test は 5000+5000 images
- label は randomize したものと正しい label を与えたものの2つの set で実験

<p align="center">
  <img src="https://imgur.com/OtT5DMh.png">
</p>

<p align="center">
  <img src="https://imgur.com/ij6IGM7.png">
</p>

正則化項の係数のサイズが適切であれば、正しいラベルで学習した場合の bayesian evidence は 0 を下回る結果を示し、モデルが確かに尤もらしくラベルを付与できていることが分かる。

**Comment by yoheikikuta** [comment]
実験: NN での generalization gap

- 800 hidden units, ReLU
- SGD w/ momentum 0.9
- learning rate 1.0
- training 1000 images
- 30 images でバッチを構成する場合と full batch でやる場合を比較

<p align="center">
  <img src="https://imgur.com/q4CkUWn.png">
</p>

small batch の方が時間は掛かるが性能は良い。これが generalization gap である。
batch size を色々変えてみたらどうなるのか？それを確認したのが以下の実験。

<p align="center">
  <img src="https://imgur.com/ZDExBfQ.png">
</p>

あまりにバッチサイズが小さいと学習できない。また、バッチサイズで汎化性能も変化する。

**Comment by yoheikikuta** [comment]
上の現象は bayesian evidence  の深さと幅の trade off であったが、これをもう少し系統的に調べる。
使う道具は SGD を連続化した確率微分方程式である。

まずは mini-batch での更新式（ $` (\epsilon / B) \sum_i \ (d C_i / dt ) `$ ）を書き換えることを考える。
上の例で、full batch と mini-batch の差が重要であったことを考えると、$` d \hat{C} / dt = (ε N/ B) \sum_i (d C_i / dt) `$ を導入して、以下のように書ける。

<p align="center">
  <img src="https://imgur.com/AxQNE2X.png">
</p>

んでこの差分の項を α と書くことにする。これまでの考察からこの差分の期待値が重要な役割を果たすことが期待される。この議論をするために、確率微分方程式（overdamped Langevin eq.）と比較することを考えるが、そのためにまずは期待値を計算しておく。

<p align="center">
  <img src="https://imgur.com/83JKKCM.png">
</p>

これを使うことで、$` <α> = 0, <α^2> = N(N/B - 1) F(ω) `$ が示せる。

**Comment by yoheikikuta** [comment]
対応付けられる確率微分方程式（overdamped Langevin eq.）を考えよう。η はノイズ。

<p align="center">
  <img src="https://imgur.com/8ZCeiPJ.png">
</p>

この η は期待値が $` <η>=0, <η(t)η(t')> = g F(ω) δ(t-t') `$ となる。ω 依存性のある項は F(ω) で記述されていて、全体的なノイズのスケールは g で規定される。

Δω の式を $` ε/N `$ が十分小さく連続近似ができると考えて $` \int_0^{ε/N} \ dω/dt dt `$ で書き、ここの確率微分方程式を代入して両辺自乗して期待値を取ることで、次の式が得られる。この計算は少し真面目に手を動かせばすぐに分かる。

<p align="center">
  <img src="https://imgur.com/LZenPSt.png">
</p>

<p align="center">
  <img src="https://imgur.com/CFaQExY.png">
</p>

この g が "noise scale" と呼んでいるもので、解の振る舞いに影響を与えるものとなっている。イメージとしてはこのノイズスケールが適切な値であれば、解の探索が良い感じに fluctuate して broad minima に辿り着けるだろう、という感じか。

**Comment by yoheikikuta** [comment]
この g が適切な値であれば汎化性能は変わらない、ということを示したのが以下の実験である。
学習率が大きすぎなければ（この場合連続近似が壊れる）、見事に解の汎化性を記述していることになっている。これは感動するレベルだ！

<p align="center">
  <img src="https://imgur.com/GsDieub.png">
</p>

MNISTの実験なので簡単だからでは？という疑問が生じるかもしれないが、 [ImageNet を1時間で学習する論文](https://arxiv.org/abs/1706.02677) でも見事にこの性質が反映されていることが見て取れる。

**Comment by yoheikikuta** [comment]
もちろん SGD の枠組みでの話ではあるが、これはとても面白かった。こういうの読むのは楽しいな。
optimal な mini-batch size とは、というのは一つの hot topic だったので、この辺の理解で結構決着が着いたのではなかろうか。

ちなみに momentum を入れた場合も幾分テクニカルではあるが同様の議論を経て、 $` g = ε N / B(1-m) `$ という関係式が得られることが分かっている。
